scipy/sparse/_index.py

```diff
@@ -32,74 +26,124 @@ class IndexMixin:
+                raise IndexError('index results in >2 dimensions')
+        else:
             if isinstance(col, INT_TYPES):
-                return self._get_arrayXint(row, col)
+                res = self._get_arrayXint(row, col)
             elif isinstance(col, slice):
-                raise IndexError('index results in >2 dimensions')
-            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
-                # special case for outer indexing
-                return self._get_columnXarray(row[:,0], col.ravel())
-
-        # The only remaining case is inner (fancy) indexing
-        row, col = _broadcast_arrays(row, col)
-        if row.shape != col.shape:
-            raise IndexError('number of row and column indices differ')
-        if row.size == 0:
-            return self.__class__(np.atleast_2d(row).shape, dtype=self.dtype)
-        return self._get_arrayXarray(row, col)
+                res = self._get_arrayXslice(row, col)
+            # arrayXarray preprocess
+            elif (row.ndim == 2 and row.shape[1] == 1
+                and (col.ndim == 1 or col.shape[0] == 1)):
+                # outer indexing
+                res = self._get_columnXarray(row[:, 0], col.ravel())
+            else:
+                # inner indexing
+                row, col = _broadcast_arrays(row, col)
+                if row.shape != col.shape:
+                    raise IndexError('number of row and column indices differ')
+                if row.size == 0:
+                    res = self.__class__(np.atleast_2d(row).shape, dtype=self.dtype)
+                else:
+                    res = self._get_arrayXarray(row, col)
+
+        # package the result and return
+        if isinstance(self, sparray) and res.shape != new_shape:
+            # handle formats that support indexing but not 1D (lil for now)
+            if self.format == "lil" and len(new_shape) != 2:
+                return res.tocoo().reshape(new_shape)
+            return res.reshape(new_shape)
+        return res
 
     def __setitem__(self, key, x):
-        row, col = self._validate_indices(key)
+        index, _ = self._validate_indices(key)
+
+        # 1D array
+        if len(index) == 1:
+            idx = index[0]
+
+            if issparse(x):
+                x = x.toarray()
+            else:
+                x = np.asarray(x, dtype=self.dtype)
+
+            if isinstance(idx, INT_TYPES):
+                if x.size != 1:
+                    raise ValueError('Trying to assign a sequence to an item')
+                self._set_int(idx, x.flat[0])
+                return
+
+            if isinstance(idx, slice):
+                # Note: Python `range` does not use lots of memory
+                idx_range = range(*idx.indices(self.shape[0]))
+                N = len(idx_range)
+                if N == 1 and x.size == 1:
+                    self._set_int(idx_range[0], x.flat[0])
+                idx = np.arange(*idx.indices(self.shape[0]))
```

Comment:
This does densify the range above; is that OK?

Reply:
I was basing this idiom on the 2D case from this module, and `get_arrayXslice` in _csr.py.  I think the comment above isn't clear. We use a Python `range` to check for the special case when the slice has a single element -- and we call `_set_int` instead of `_set_array`.  Part of the lack of clarity is that this shortcut should be finished with a `return` and it isn't. So, in fact the code in the PR was setting the element once in `_set_int` and then again in `_set_array` later on.  That's my mistake. And the tests don't catch it because it is just doing extra work -- not creating a wrong answer.

I have added a comment and a return statement to make this more clear, and actually take the shortcut when we can.